nexus/reconfigurator/planning/src/blueprint_builder/builder.rs

```diff
@@ -1098,16 +1076,97 @@ impl<'a> BlueprintBuilder<'a> {
                 .map_err(|err| Error::SledEditError { sled_id, err })?;
         }
 
-        // Remove any disks that appear in the blueprint, but not the database
-        for disk_id in blueprint_disk_ids {
-            if !database_disk_ids.contains(&disk_id) {
+        let final_counts = editor.edit_counts();
+        Ok(final_counts.difference_since(initial_counts))
+    }
+
+    /// Set the state of the disk to `Decommissioned` for any expunged disks if:
+    ///
+    /// (1) the sled where the disk resides has been expunged.
+    ///
+    ///   or
+    ///
+    /// (2) the sled-agent where the disk resides has seen the
+    /// request to expunge the disk as reflected in inventory. We use
+    /// `BlueprintPhysicalDisksConfig::generation` from the parent blueprint and
+    /// `SledAgent::omicron_physical_disks_generation` from inventory to see how
+    /// up to date the inventory is with respect to the parent_blueprint.
+    ///
+    ///
+    /// Caveats:
+    ///
+    ///   When the disk resides on an expunged sled, we can immediately
+    ///   decommission it, even if it has had its disposition set to expunged in
+    ///   *this* planner pass.
+    ///
+    ///   However, if the sled is still in service, we must wait for any
+    ///   disks expunged in this planner pass to be seen by the sled-agent and
+    ///   reflected in inventory before we can decommission them. Since any
+    ///   modifications made in this planner pass have not yet been sent to the
+    ///   sled-agents, we only check for expunged sleds in the parent_blueprint.
+    ///
+    ///
+    /// Called by the planner in the `do_plan_decommission()` stage of planning
```

Comment:
Should this method be in the planner instead? "Compare inventory and make changes" vaguely feels to me like it's a planner thing, and the builder should only expose whatever's needed for the "make changes" bit.

If that seems right, I think I'd make two suggestions:

* Move the "decommission disks for expunged sleds" logic into `BlueprintBuilder::expunge_sled()`, which already does everything else needed for expunging an entire sled.
* Add a `BlueprintBuilder::decommision_disks()` method that the planner could call in the `seen_generation >= as_of_generation` case.

(This mostly comes from comparing this branch against my first draft of doing all this same work for zones.)

Reply:
Very good point. I think I was cribbing too much off of names like `sled_add_disks` while not fully considering the planner builder split. Will do.

Reply:
Done in [2531105](https://github.com/oxidecomputer/omicron/pull/7286/commits/2531105c56870e196236ad82cc9aad24829c148f).

The code is much clearer, thanks!